File: elixir/beer-song/beer.exs
defmodule Beer do

  def sing(start, finish // 0) do
    song = gather_verses(start, finish) |> combine_verses 
    Enum.join [song, "\n"]
  end

  def verse(0) do
    "#{String.capitalize(beer_text(0))} on the wall, #{beer_text(0)}.\nGo to the store and buy some more, #{beer_text(99)} on the wall.\n"
  end

  def verse(1) do
    "#{beer_text(1)} on the wall, #{beer_text(1)}.\nTake it down and pass it around, #{beer_text(0)} on the wall.\n"
  end

  def verse(existing) do
    remaining = existing - 1
    "#{beer_text(existing)} on the wall, #{beer_text(existing)}.\nTake one down and pass it around, #{beer_text(remaining)} on the wall.\n"
  end

  defp combine_verses(verses) do
    Enum.join verses, "\n"
  end

  defp gather_verses(start, finish) do
    Enum.map(start..finish, fn (x) -> verse x end ) 
  end

  defp beer_text(0) do
    "no more bottles of beer"
  end

  defp beer_text(1) do
    "1 bottle of beer"
  end

  defp beer_text(x) do
    "#{x} bottles of beer"
  end
end

File: javascript/beer-song/beer.js
var VerseText = {
  bottleText: function (count) { 
    return (count == 1) ? "bottle" : "bottles";
  },
  verseX: function (count) { 
    var remaining = count - 1;
    return "" + count + " bottles of beer on the wall, " + count + " bottles of beer.\nTake one down and pass it around, " + remaining + " " + VerseText.bottleText(remaining) + " of beer on the wall.\n";
  }, 
  verse1: function (){ 
    return "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n";
  },
  verse0: function (){ 
    return "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n";
  }
};

var Beer = {
  verse: function (count) { 
    if (count > 1) { 
      return VerseText.verseX(count);
    } else if (count === 1) { 
      return VerseText.verse1();
    } else { 
      return VerseText.verse0();
    }
  },

  sing: function (start, end) { 
    if (!end) { end = 0; }
    var lines = [];
    for ( var i = start; i >= end; i-- ) {
      lines.push( Beer.verse(i));
    }
    return lines.join("\n");
  }
};

module.exports = Beer;
